lms/services/canvas_api/_authenticated.py

```diff
@@ -57,16 +57,31 @@ def send(self, method, path, schema, params=None):
         call_args = (method, path, schema, params)
 
         try:
-            auth_header = f"Bearer {self._oauth2_token_service.get().access_token}"
-            return self._client.send(*call_args, headers={"Authorization": auth_header})
+            oauth2_token = self._oauth2_token_service.get()
+        except NoOAuth2Token as err:
+            raise CanvasAPIAccessTokenError(
+                explanation="We don't have a Canvas API access token for this user",
+                response=None,
+            ) from err
 
-        except CanvasAPIAccessTokenError:
-            refresh_token = self._oauth2_token_service.get().refresh_token
+        access_token = oauth2_token.access_token
+        refresh_token = oauth2_token.refresh_token
+
+        try:
+            return self._client.send(
+                *call_args,
+                headers={"Authorization": f"Bearer {access_token}"},
+            )
+        except CanvasAPIAccessTokenError as err:
             if not refresh_token:
                 raise
 
-            auth_header = f"Bearer {self.get_refreshed_token(refresh_token)}"
-            return self._client.send(*call_args, headers={"Authorization": auth_header})
+            new_access_token = self.get_refreshed_token(refresh_token)
+
+            return self._client.send(
+                *call_args,
+                headers={"Authorization": f"Bearer {new_access_token}"},
+            )
```

Comment:
I rewrote this method a little to make it clearer what it does:

1. It gets the `OAuth2Token` from the DB
2. If there's no `OAuth2Token` in the DB (`NoOAuth2Token`) it raises `CanvasAPIAccessTokenError`
3. It sends the request
4. If the request raises `CanvasAPIAccessTokenError` (e.g. our access token was deleted from Canvas, or is expired) it:
   1. Just raises the `CanvasAPIAccessTokenError` if we don't have a refresh token
   2. Gets a new access token if we do have a refresh token
   3. Re-sends the request with the new access token

Reply:
Not sure it makes much difference but we could out-dent the stuff after the raise to take advantage of the fast return for less nested code.

Reply:
The more I look at `raise_from` the less I like it. This is accepting a soup of different types and duck typing the lot. I think it would be easier to pick apart if we could specify what types we are expecting, so a few more `if isinstance(...` and then pull bits off.

Reply:
Yeah, this is dumb. It's only called in three places all inside _send_prepared. I think we're taking disparate situations, crushing them together, and then pulling them back apart again needlessly.

I think this might be worth a few refactoring PRs to try and knock back into a more sensible shape. I suspect a lot of this code would just dissappear. Some ideas:

 * Stop the duck typing stuff:
    * We are actually interested in `messages` if it's a `ValidationError`
    * `response` if it's a `pyramid.HTTPException` (which `ValidationError` is too)
 * Move the decision of what exception to raise into the basic client
 * Once the Canvas specific stuff is out move the logic for to `ExternalRequestError`